### ISSUE TITLE
Cached target files are all -1

### ISSUE THREAD

**Issue by awebson** [bigscience-workshop/promptsource#455]
Currently, these target files have all -1. Colin proposed a fix in Slack and I will try that first thing tomorrow morning.

```
######
winogrande_winogrande_xl
Prompt: _Replace
gs://bigscience/experiment_d/finetune-t5-xxl-lm-d4-091621-512/validation_eval/winogrande_winogrande_xl_Replace_targets
Prompt: _does_underscore_refer_to
gs://bigscience/experiment_d/finetune-t5-xxl-lm-d4-091621-512/validation_eval/winogrande_winogrande_xl_does_underscore_refer_to_targets
Prompt: _fill_in_the_blank
gs://bigscience/experiment_d/finetune-t5-xxl-lm-d4-091621-512/validation_eval/winogrande_winogrande_xl_fill_in_the_blank_targets
Prompt: _stand_for
gs://bigscience/experiment_d/finetune-t5-xxl-lm-d4-091621-512/validation_eval/winogrande_winogrande_xl_stand_for_targets
Prompt: _underscore_refer_to
gs://bigscience/experiment_d/finetune-t5-xxl-lm-d4-091621-512/validation_eval/winogrande_winogrande_xl_underscore_refer_to_targets
######
winogrande_winogrande_debiased
Prompt: _Replace
gs://bigscience/experiment_d/finetune-t5-xxl-lm-d4-091621-512/validation_eval/winogrande_winogrande_debiased_Replace_targets
Prompt: _does_underscore_refer_to
gs://bigscience/experiment_d/finetune-t5-xxl-lm-d4-091621-512/validation_eval/winogrande_winogrande_debiased_does_underscore_refer_to_targets
Prompt: _fill_in_the_blank
gs://bigscience/experiment_d/finetune-t5-xxl-lm-d4-091621-512/validation_eval/winogrande_winogrande_debiased_fill_in_the_blank_targets
Prompt: _stand_for
gs://bigscience/experiment_d/finetune-t5-xxl-lm-d4-091621-512/validation_eval/winogrande_winogrande_debiased_stand_for_targets
Prompt: _underscore_refer_to
gs://bigscience/experiment_d/finetune-t5-xxl-lm-d4-091621-512/validation_eval/winogrande_winogrande_debiased_underscore_refer_to_targets
######
race_high
Prompt: _Read_the_article_and_answer_the_question_no_option_
gs://bigscience/experiment_d/finetune-t5-xxl-lm-d4-091621-512/validation_eval/race_high_Read_the_article_and_answer_the_question_no_option__targets
Prompt: _Select_the_best_answer_generate_span_
gs://bigscience/experiment_d/finetune-t5-xxl-lm-d4-091621-512/validation_eval/race_high_Select_the_best_answer_generate_span__targets
######
race_middle
Prompt: _Read_the_article_and_answer_the_question_no_option_
gs://bigscience/experiment_d/finetune-t5-xxl-lm-d4-091621-512/validation_eval/race_middle_Read_the_article_and_answer_the_question_no_option__targets
Prompt: _Select_the_best_answer_generate_span_
gs://bigscience/experiment_d/finetune-t5-xxl-lm-d4-091621-512/validation_eval/race_middle_Select_the_best_answer_generate_span__targets
######
social_i_qa
Prompt: _Generate_answer
gs://bigscience/experiment_d/finetune-t5-xxl-lm-d4-091621-512/validation_eval/social_i_qa_Generate_answer_targets
Prompt: _I_was_wondering
gs://bigscience/experiment_d/finetune-t5-xxl-lm-d4-091621-512/validation_eval/social_i_qa_I_was_wondering_targets
Prompt: _Show_choices_and_generate_answer
gs://bigscience/experiment_d/finetune-t5-xxl-lm-d4-091621-512/validation_eval/social_i_qa_Show_choices_and_generate_answer_targets
######
super_glue_copa
Prompt: _C1_or_C2_premise_so_because_
gs://bigscience/experiment_d/finetune-t5-xxl-lm-d4-091621-512/validation_eval/super_glue_copa_C1_or_C2_premise_so_because__targets
Prompt: __As_a_result_C1_or_C2_
gs://bigscience/experiment_d/finetune-t5-xxl-lm-d4-091621-512/validation_eval/super_glue_copa__As_a_result_C1_or_C2__targets
Prompt: __What_could_happen_next_C1_or_C2_
gs://bigscience/experiment_d/finetune-t5-xxl-lm-d4-091621-512/validation_eval/super_glue_copa__What_could_happen_next_C1_or_C2__targets
Prompt: __which_may_be_caused_by
gs://bigscience/experiment_d/finetune-t5-xxl-lm-d4-091621-512/validation_eval/super_glue_copa__which_may_be_caused_by_targets
Prompt: __why_C1_or_C2
gs://bigscience/experiment_d/finetune-t5-xxl-lm-d4-091621-512/validation_eval/super_glue_copa__why_C1_or_C2_targets
Prompt: _best_option
gs://bigscience/experiment_d/finetune-t5-xxl-lm-d4-091621-512/validation_eval/super_glue_copa_best_option_targets
Prompt: _cause_effect
gs://bigscience/experiment_d/finetune-t5-xxl-lm-d4-091621-512/validation_eval/super_glue_copa_cause_effect_targets
Prompt: _choose
gs://bigscience/experiment_d/finetune-t5-xxl-lm-d4-091621-512/validation_eval/super_glue_copa_choose_targets
Prompt: _exercise
gs://bigscience/experiment_d/finetune-t5-xxl-lm-d4-091621-512/validation_eval/super_glue_copa_exercise_targets
Prompt: _i_am_hesitating
gs://bigscience/experiment_d/finetune-t5-xxl-lm-d4-091621-512/validation_eval/super_glue_copa_i_am_hesitating_targets
Prompt: _more_likely
gs://bigscience/experiment_d/finetune-t5-xxl-lm-d4-091621-512/validation_eval/super_glue_copa_more_likely_targets
Prompt: _plausible_alternatives
gs://bigscience/experiment_d/finetune-t5-xxl-lm-d4-091621-512/validation_eval/super_glue_copa_plausible_alternatives_targets
######
piqa
Prompt: _finish_sentence_with_correct_choice
gs://bigscience/experiment_d/finetune-t5-xxl-lm-d4-091621-512/validation_eval/piqa_finish_sentence_with_correct_choice_targets
Prompt: _pick_correct_choice_with_choice_given_before_goal
gs://bigscience/experiment_d/finetune-t5-xxl-lm-d4-091621-512/validation_eval/piqa_pick_correct_choice_with_choice_given_before_goal_targets
Prompt: _what_is_the_correct_ending
gs://bigscience/experiment_d/finetune-t5-xxl-lm-d4-091621-512/validation_eval/piqa_what_is_the_correct_ending_targets
######
story_cloze_2016
Prompt: _Answer_Given_options
gs://bigscience/experiment_d/finetune-t5-xxl-lm-d4-091621-512/validation_eval/story_cloze_2016_Answer_Given_options_targets
Prompt: _Choose_Story_Ending
gs://bigscience/experiment_d/finetune-t5-xxl-lm-d4-091621-512/validation_eval/story_cloze_2016_Choose_Story_Ending_targets
Prompt: _Movie_What_Happens_Next
gs://bigscience/experiment_d/finetune-t5-xxl-lm-d4-091621-512/validation_eval/story_cloze_2016_Movie_What_Happens_Next_targets
Prompt: _Novel_Correct_Ending
gs://bigscience/experiment_d/finetune-t5-xxl-lm-d4-091621-512/validation_eval/story_cloze_2016_Novel_Correct_Ending_targets
Prompt: _Story_Continuation_and_Options
gs://bigscience/experiment_d/finetune-t5-xxl-lm-d4-091621-512/validation_eval/story_cloze_2016_Story_Continuation_and_Options_targets
######
hellaswag
Prompt: _Randomized_prompts_template
gs://bigscience/experiment_d/finetune-t5-xxl-lm-d4-091621-512/validation_eval/hellaswag_Randomized_prompts_template_targets
Prompt: _complete_first_then
gs://bigscience/experiment_d/finetune-t5-xxl-lm-d4-091621-512/validation_eval/hellaswag_complete_first_then_targets
```

**Comment by stephenbach** [comment]
I'm not sure that this is related to whitespace in `get_fixed_answer_choices_list()`. First, we strip whitespace from the elements:
https://github.com/bigscience-workshop/promptsource/blob/main/promptsource/templates.py#L164
Second, at least some templates e.g., winogrande_does_underscore_refer_to doesn't have fixed answer choices anyway, so this method returns None

**Comment by craffel** [comment]
For context/posterity, this is what I posted in Slack:

> Looked into this. My guess is that what's happening is that template.get_fixed_answer_choices_list is returning a list of answer choices here:
https://github.com/bigscience-workshop/promptsource/blob/b491ab93b9bdd8f1e0d6069f3c2956fa83c5ee7d/promptsource/seqio_tasks/tasks.py#L45
so the string_label_to_class_id postprocessor is getting applied to the task here:
https://github.com/bigscience-workshop/promptsource/blob/b491ab93b9bdd8f1e0d6069f3c2956fa83c5ee7d/promptsource/seqio_tasks/tasks.py#L49
but the targets don't actually match anything in the list returned by template.get_fixed_answer_choices_list . My best guess as to why is that the targets get whitespace-stripped before the  string_label_to_class_id postprocessor  is applied, but the entries of the list returned by template.get_fixed_answer_choices_list don't get whitespace-stripped. I think we probably just need to change
label_classes=template.get_fixed_answer_choices_list() to label_classes=[l.strip() for l in template.get_fixed_answer_choices_list()] here:
https://github.com/bigscience-workshop/promptsource/blob/b491ab93b9bdd8f1e0d6069f3c2956fa83c5ee7d/promptsource/seqio_tasks/tasks.py#L50

But, @stephenbach , I see that you're right about stripping in get_fixed_answer_choices_list. So, I think my guess that this has to do with stripping is probably wrong - but I still would be reasonably sure this has to do with a mismatch between the actual targets that get produced and the list returned by get_fixed_answer_choices_list, since that's the way that string_label_to_class_id produces -1: https://github.com/google-research/text-to-text-transfer-transformer/blob/ebd6ce98385ac8259119f005d404c51d42a02349/t5/data/postprocessors.py#L41

**Comment by stephenbach** [comment]
Do any of these templates have fixed answer choices? All the ones I've checked so far have per-example choices. Is the issue that this evaluation should use `get_answer_choices_list(example)`?
https://github.com/bigscience-workshop/promptsource/blob/main/promptsource/templates.py#L134

**Comment by awebson** [comment]
I think Steve is right. I looked into this a bit and the fixed choices do not have superfluous whitespace. And per-example choices really should call `get_answer_choices_list(example)`, which is currently not implemented in tasks.py. 

@stephenbach What exactly is the format of `example` that `get_answer_choices_list()` accepts? Just the raw string of a data example? 

@craffel I think some seqio pre/post-processor has to, for each example, give promptsource the example and receive an example-specific list of answer choices for rank eval. Sadly I don't really know enough about the plumbing here to fix or test it, so I'm afraid that we will have to rely on you to save the day.

**Comment by stephenbach** [comment]
It’s expecting a dictionary like from the datasets library, but this can be changed if helpful.

**Comment by craffel** [comment]
I'm confused. If a dataset has per-example answer choices, shouldn't `template.get_fixed_answer_choices_list()` return None? And therefore the `t5.data.postprocessors.string_label_to_class_id` postprocessor, which is the source of these -1s, would not be applied? https://github.com/bigscience-workshop/promptsource/blob/main/promptsource/seqio_tasks/tasks.py#L45?

**Comment by awebson** [comment]
Not sure why but, for example, calling `template.get_fixed_answer_choices_list()` on Winogrande’s "does underscore refer to" prompt returns `['', '']` which of course evaluates to True in Python.

**Comment by stephenbach** [comment]
Ok, that's the issue. It's not correctly detecting that the answer choices depend on the example. Will get PR posted asap.

**Comment by stephenbach** [comment]
Potential fix in #467.
